Query metric per host 1155

alourie opened the pull request:
Attempt to solve #1155 

This ended up being a bit bigger than initially expected.

Nevertheless, the majority of work was to update the tests to fetch the correct host they're doing the work for and sending it over as a parameter to other functions. Additionally, it should be somewhat easier to add new metrics to the query execution.

The tests seem to pass successfully now.

@Zariel would love your feedback on this.

annismckenzie commented on session.go:
```diff
@@ -658,6 +658,11 @@ func (s *Session) connect(host *HostInfo, errorHandler ConnErrorHandler) (*Conn,
 }
 
+type queryMetric struct {
+	attempts     int
```
Make those `uint64` because there's no good reason why it should ever be negative.

alourie replied:
agreed.

annismckenzie commented on session.go:
```diff
@@ -698,9 +702,22 @@ func (q *Query) defaultsFromSession() {
+	defer s.pool.mu.RUnlock()
+
+	// Range over all defined IPs and prebuild metrics map
+	for _, hostConn := range s.pool.hostConnPools {
```
Is there a possibility that a host is added to the pool after this is done? I'm pretty sure that can happen if you add a new node to the cluster and it's gossiped around the cluster through the control connection. In that case you might panic in the map accesses below.

alourie replied:
as mentioned above, will probably remove this completely.

annismckenzie commented on conn_test.go:
```diff
@@ -690,8 +773,9 @@ func TestFrameHeaderObserver(t *testing.T) {
-func NewTestServer(t testing.TB, protocol uint8, ctx context.Context) *TestServer {
-	laddr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
+func NewTestServerWithAddress(addr string, t testing.TB, protocol uint8, ctx context.Context) *TestServer {
+
```
drop this empty line

annismckenzie commented on conn_test.go:
```diff
@@ -721,6 +805,11 @@ func NewTestServer(t testing.TB, protocol uint8, ctx context.Context) *TestServe
 	go srv.serve()
 
 	return srv
+
```
drop this empty line

annismckenzie commented on session.go:
```diff
@@ -808,9 +833,10 @@ func (q *Query) execute(conn *Conn) *Iter {
-	q.attempts++
-	q.totalLatency += end.Sub(start).Nanoseconds()
-	// TODO: track latencies per host and things as well instead of just total
+
```
drop this empty line

annismckenzie commented on session.go:
```diff
@@ -1525,9 +1557,11 @@ func (b *Batch) WithTimestamp(timestamp int64) *Batch {
-	b.attempts++
-	b.totalLatency += end.Sub(start).Nanoseconds()
-	// TODO: track latencies per host and things as well instead of just total
+
```
drop this empty line

annismckenzie commented on session.go:
```diff
@@ -1414,7 +1441,9 @@ func (s *Session) NewBatch(typ BatchType) *Batch {
 		keyspace:         s.cfg.Keyspace,
+		metrics:          initializeMetrics(s),
 	}
+
```
drop this empty line

alourie commented on session_test.go:
```diff
@@ -191,25 +203,29 @@ func TestBatchBasicAPI(t *testing.T) {
-		t.Fatalf("expceted batch.Attempts() to return %v, got %v", 1, b.Attempts())
+	b.metrics[ip] = &queryMetric{attempts: 1}
+	if b.Attempts(host) != 1 {
+		t.Fatalf("expceted batch.Attempts() to return %v, got %v", 1, b.Attempts(host))
```
sure :-)

alourie commented on session.go:
```diff
@@ -698,9 +702,22 @@ func (q *Query) defaultsFromSession() {
 	s.mu.RUnlock()
 }
 
+func initializeMetrics(s *Session) map[string]*queryMetric {
```
the main reason is that Query and Batch initialisers are the same an don't really need to work on a query object.

On the other hand, acking the comment on the initialiser itself, I will probably get rid of it completely as it would no longer make sense if the nodes are added later.

alourie commented:
@annismckenzie more than OK, I really appreciate it!

alourie commented:
@annismckenzie I've updated the PR, would love if you could have a look. Thanks.

annismckenzie commented on session.go:
```diff
@@ -698,9 +702,20 @@ func (q *Query) defaultsFromSession() {
 }
 
+func (q *Query) getHostMetrics(host *HostInfo) *queryMetrics {
+	_, exists := q.metrics[host.connectAddress.String()]
```
Only look it up once in this method?

alourie replied:
Considering that this call is used for both a lookup _and_ initiating a zero value, I really can't think of a more "efficient" or better way to do it, except splitting it into 2 functions, 1 to initiate the map and the other is to fetch. Personally, don't think it makes it all that better, but is that what you're thinking of?

annismckenzie replied:
I meant something like this:
```go
func (q *Query) getHostMetrics(host *HostInfo) *queryMetrics {
	metrics, exists := q.metrics[host.connectAddress.String()]
	if !exists {
		// if the host is not in the map, it means it's been accessed for the first time
		metrics = &queryMetrics{attempts: 0, totalLatency: 0}
		q.metrics[host.connectAddress.String()] = metrics
	}
 	return metrics
}
```

alourie replied:
yea, thought about this one too, but didn't have a good opinion on it... would you think it looks better? If so, can easily change.

annismckenzie replied:
I think it does look better. 🙈

alourie replied:
Done

annismckenzie commented on session.go:
```diff
@@ -1431,14 +1467,16 @@ func (b *Batch) Keyspace() string {
-	if b.attempts > 0 {
-		return b.totalLatency / int64(b.attempts)
+func (b *Batch) Latency(host *HostInfo) int64 {
+	hostMetric := b.getHostMetrics(host)
```
Why doesn't the `Batch` have a mutex while the `Query` still does? 🤔

alourie replied:
They both don't have a mutex. The only mutexes in Query (but also in Batches) are the ones embedded in Session, where there are 2: one for the session itself and the second is for the hostPool access. Am I missing something? :-)

annismckenzie replied:
Don't know what I was seeing – weird. Diregard. 😬

Zariel commented:
@alourie needs a rebase now

alourie commented:
@Zariel @annismckenzie okay....So I've rebased with the last master, cleaned up the code a bit, removed yet unnecessary locks. Also found a small bug in the query_executor, which I also fixed, added a separate test case for that for future reference and finally updated the testRetryPolicy to be a bit more flexible.

Travis is now happy, would love your feedback/approval ;-)

annismckenzie commented on session.go:
```diff
@@ -713,14 +729,18 @@ func (q Query) String() string {
 //Attempts returns the number of times the query was executed.
-func (q *Query) Attempts() int {
-	return q.attempts
+func (q *Query) Attempts(host *HostInfo) uint64 {
```
The main gripe I have with this approach is that this a) break backwards compatibility in a major way and b) now produces a different behavior even with the default `SimpleRetryPolicy`. The metrics are tracked by host and default retry type of the `SimpleRetryPolicy` is to `RetryNextHost` which means the attempts are now also _per host_. The max retries are ~therefor also `number of hosts * max retries - (number of hosts - 1)` (don't ask, the gist: it fills~ (that's not correct) not honored because on each successive retry (on the next host) the attempts for that will be 0 and it will exhaust the host pool and fail with a different error than before. Mind you, I'm not harking on the general direction or the feature (it _is_ very useful) but maybe leave the current signatures alone and keep track of a summed version of attempts and latencies (as it is right now) and make the per host latencies and attempts available to the retry policy through some other means so going forward one can create an advanced retry policy that takes those into account? 🤔

alourie replied:
Hmmm...now I'm kinda confused.

So, you're right, the current `queryExecution` only checks that the total accumulated attempts across all hosts have not surpassed the number set in the retry policy, and this approach changes that. I was wrong in my understanding of what it does.

So...I think I understand it now. What I will probably do is introduce a way to find the total count by summing the map values; also will see how to keep signatures the same.

Thanks for the review!

alourie commented on query_executor.go:
```diff
@@ -45,9 +45,11 @@ func (q *queryExecutor) checkRetryPolicy(rq ExecutableQuery, err error) (RetryTy
 	}
+
+	// p.Attempt is false, should not retry anything
 	return p.GetRetryType(err), err
```
Well, that's also a problem. If the retryPolicy is implemented similarly to the new testRetryPolicy, then when those attempts are done it returns `RetryNextHost`, not Rethrow. If you ignore that, then you're throwing an error instead of trying the next host.

That's why I ended up accepting the need to call it here, but check it in the queryExecutor. I don't like both solutions, but at least the current one seems to be working correctly.

annismckenzie replied:
Would you mind if I fix this in my branch because I'm working on #1161 and it touches the same code path again in the query executor?

alourie commented on session.go:
```diff
@@ -826,8 +846,8 @@ func (q *Query) attempt(keyspace string, end, start time.Time, iter *Iter, host
 			End:       end,
 			Rows:      iter.numRows,
 			Host:      host,
+			Metrics:   q.metrics[host.ConnectAddress().String()],
```
you're right, don't know what I was thinking...will fix

alourie commented:
Totally, go ahead. I planned to look at #1161 if you can't ;-)

On Fri., 24 Aug. 2018, 07:28 Daniel Lohse, <notifications@github.com> wrote:

> *@annismckenzie* commented on this pull request.
> ------------------------------
>
> In query_executor.go
> <https://github.com/gocql/gocql/pull/1156#discussion_r212470304>:
>
> >  		return p.GetRetryType(err), nil
>  	}
> +
> +	// p.Attempt is false, should not retry anything
>  	return p.GetRetryType(err), err
>
> Would you mind if I fix this in my branch because I'm working on #1161
> <https://github.com/gocql/gocql/issues/1161> and it touches the same code
> path again in the query executor?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gocql/gocql/pull/1156#discussion_r212470304>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADdBxw7B_ZiZh7gmNB4zAIlCw-sqi8yks5uTyV2gaJpZM4WBvE0>
> .
>
-- 
Best Regards,
Alex Lourie

annismckenzie commented:
Nah, already deep in the weeds on #1161. I'll count on your :eyes: though. 😬

annismckenzie commented on conn_test.go:
```diff
@@ -299,15 +308,66 @@ func (s *testRetryPolicy) AttemptTimeout() time.Duration {
+		Logger.Printf("Observed query %q. Returned %v rows, took %v on host %q. Error: %q\n", q.Statement, q.Rows, q.End.Sub(q.Start), host, q.Err)
+	}
+
+	// Uncomment and add verbosity?
```
Drop and move the missing log entries `q.Metrics.attempts` and `q.Metrics.totalLatency` into the log message above.

annismckenzie commented on control.go:
```diff
@@ -453,7 +453,11 @@ func (c *controlConn) query(statement string, values ...interface{}) (iter *Iter
 		}
 
-		q.attempts++
+		host := c.getConn().host
```
Collapse this down to:
```go
c.session.mu.Lock()
q.getHostMetrics(c.getConn().host).attempts++
c.session.mu.Unlock()
```

It's just a lot of variables for no purpose. You can leave the `host :=` line but I'd definitely collapse the attempt increase because it reads better.

alourie replied:
I'm not sure this one follows the regular query execution. It seems we're constructing and executing the statement directly on conn, and not via the reqular query mechanisms. The retry policy is not constructed, which I reckon means it would be set to `SimpleRetryPolicy{numAtt: 3}`...so we still need to count attempts, but probably not locking.... :-)

annismckenzie commented on conn_test.go:
```diff
@@ -299,15 +308,66 @@ func (s *testRetryPolicy) AttemptTimeout() time.Duration {
+	// 	q.Statement, q.Rows, q.End.Sub(q.Start), host, q.Metrics.attempts, q.Metrics.totalLatency, q.Err)
+}
+
+func (o *testQueryObserver) GetMetrics(host *HostInfo) *queryMetrics {
```
no need for an exported function in a test

alourie replied:
indeed

annismckenzie commented on conn_test.go:
```diff
@@ -57,13 +58,17 @@ func TestJoinHostPort(t *testing.T) {
 
-func testCluster(addr string, proto protoVersion) *ClusterConfig {
-	cluster := NewCluster(addr)
+func testMultiNodeCluster(addresses []string, proto protoVersion) *ClusterConfig {
```
Nice, I'll need this too. :+1:

annismckenzie commented on policies_test.go:
```diff
@@ -358,6 +365,7 @@ func TestDowngradingConsistencyRetryPolicy(t *testing.T) {
 		if !c.allow && rt.Attempt(q) {
 			t.Fatalf("should not allow retry after %d attempts", c.attempts)
 		}
+
```
drop

alourie replied:
daymn, missed this

annismckenzie commented on query_executor.go:
```diff
@@ -48,7 +48,8 @@ func (q *queryExecutor) checkRetryPolicy(rq ExecutableQuery, err error) (RetryTy
 	if p.Attempt(rq) {
 		return p.GetRetryType(err), nil
 	}
-	return p.GetRetryType(err), err
```
Please revert this. I'm opening a separate PR for this (will include a test).

annismckenzie replied:
Opened #1164 for this.

annismckenzie commented on session.go:
```diff
@@ -673,14 +678,13 @@ type Query struct {
 	session               *Session
 	rt                    RetryPolicy
 	binding               func(q *QueryInfo) ([]interface{}, error)
-	attempts              int
```
I'd keep this and rename it to `totalAttempts`. I'll follow this up below.

annismckenzie commented on session.go:
```diff
@@ -699,9 +703,23 @@ func (q *Query) defaultsFromSession() {
+		hostMetrics = &queryMetrics{attempts: 0, totalLatency: 0}
+		q.metrics[host.ConnectAddress().String()] = hostMetrics
+	}
+
```
drop

annismckenzie commented on session.go:
```diff
@@ -814,9 +845,11 @@ func (q *Query) execute(conn *Conn) *Iter {
+	hostMetrics := q.getHostMetrics(host)
+	q.session.mu.Lock()
+	hostMetrics.attempts++
+	hostMetrics.totalLatency += end.Sub(start).Nanoseconds()
```
I'd keep track of the total number of attempts and total latency in here (as per my recommendation to keep the fields on `Query` above) and also calculate the average as there's already an attempt, thereby you can do away with the check for 0 division. That way, you can get rid of the lock dance in both `Latency()` and `Attempts()` as well as the for loops and so on.

annismckenzie replied:
That'll simplify the code above and because you're protecting the recording of the attempt with the lock anyway it's easy track of these values without fear they'll drift.

annismckenzie commented on session.go:
```diff
@@ -673,14 +678,13 @@ type Query struct {
 	rt                    RetryPolicy
 	binding               func(q *QueryInfo) ([]interface{}, error)
-	attempts              int
-	totalLatency          int64
```
add `averageLatency int`

annismckenzie commented on session.go:
```diff
@@ -1557,8 +1619,8 @@ func (b *Batch) attempt(keyspace string, end, start time.Time, iter *Iter, host
 		Host:    host,
+		Metrics: hostMetrics,
 		Err:     iter.err,
-		Attempt: b.attempts,
```
If you're good with my suggestions above, you may want to restore this field and pass `b.totalAttempts`. On the other hand, query observers could be fine with calculating it from `hostMetrics`. Your call.

annismckenzie commented on session.go:
```diff
@@ -826,8 +859,8 @@ func (q *Query) attempt(keyspace string, end, start time.Time, iter *Iter, host
 			End:       end,
 			Rows:      iter.numRows,
 			Host:      host,
+			Metrics:   hostMetrics,
```
I hope it's okay that the query observer will access the fields in `hostMetrics` while they could be changed – you're protecting write access above but that mutex is unlocked in line 852. Could that race? 🤔 😅

annismckenzie commented on session.go:
```diff
@@ -699,9 +703,23 @@ func (q *Query) defaultsFromSession() {
 
+func (q *Query) getHostMetrics(host *HostInfo) *queryMetrics {
+	q.session.mu.Lock()
+	defer q.session.mu.Unlock()
```
As there's no way you're returning from this function early, I'd drop the defer and move the mutex unlock down. It's pretty clear so there's no need for the defer overhead, especially because this could be called a sizable number of times while a query is executing and the retry policy consulted. Come to think of it, making this available to the retry policy isn't implemented yet, right?

alourie replied:
no, not implemented, and I'm not sure retryPolicy needs to deal with metrics at all. But considering the Attempt() gets the whole query, it may, in fact, use the metrics if relevant for the retry policy.

annismckenzie commented on policies_test.go:
```diff
@@ -263,8 +263,12 @@ func TestSimpleRetryPolicy(t *testing.T) {
 		{5, false},
 	}
 
+	// initiate metrics map
```
I think you can do away with this comment as well as the ones in lines 269 and 354.

alourie replied:
:-)

annismckenzie commented on session.go:
```diff
@@ -1421,11 +1454,25 @@ func (s *Session) NewBatch(typ BatchType) *Batch {
 }
 
+func (b *Batch) getHostMetrics(host *HostInfo) *queryMetrics {
+	b.mu.Lock()
```
The same suggestion I had above with changing this machinery applies to the `Batch` as well. Shame that it's duplicated. This screams separate manager object that holds query attempt metrics as well as latencies and total attempts. It would also include the mutex. Sorry for bringing it up. 😇

alourie replied:
I totally agree :-) I've almost implemented it on Friday, just was too sleepy to make sure the tests pass. Hopefully it will resolve all the gripes with the current implementation.

Zariel commented on conn_test.go:
```diff
@@ -57,13 +58,17 @@ func TestJoinHostPort(t *testing.T) {
 
-func testCluster(addr string, proto protoVersion) *ClusterConfig {
-	cluster := NewCluster(addr)
+func testMultiNodeCluster(addresses []string, proto protoVersion) *ClusterConfig {
```
Define this like 

```go
func testMultiNodeCluster(proto protoVersion, addresses ...string) *ClusterConfig
```

That way callers dont need to wrap a single addr in a slice

alourie replied:
wasn't sure that it's ok to change the order, but will do now :-)

Zariel commented on session.go:
```diff
@@ -714,13 +732,26 @@ func (q Query) String() string {
 func (q *Query) Attempts() int {
-	return q.attempts
+	q.session.mu.Lock()
+	defer q.session.mu.Unlock()
```
Why are we now locking the session mutex around the place?

alourie replied:
will change

Zariel commented on session.go:
```diff
@@ -699,9 +703,23 @@ func (q *Query) defaultsFromSession() {
 }
 
+func (q *Query) getHostMetrics(host *HostInfo) *queryMetrics {
+	q.session.mu.Lock()
```
What is this lock protecting?

alourie replied:
@annismckenzie You're right, I am working on speculative queries executions, and indeed I use wait groups to keep track of the parallel query executions. The lock here is protecting from the parallel access of the attempts number in the query from different goroutines ....but mind you that I've built the speculative stuff a week before I started this work, and since I've discovered a few holes in my understanding of how retryPolicies and retries in general work. So much could change  and I might remove the locks after all.

Zariel commented on integration.sh:
```diff
@@ -72,13 +72,13 @@ function run_tests() {
 		sleep 1s
 
-		go test -tags "ccm gocql_debug" -timeout=5m -race $args
+		go test -tags "ccm gocql_debug" -timeout=15m -race $args
```
remove unrelated change

alourie replied:
of course

Zariel commented:
@alourie this is a -1 from due to the lock required on the session, I don't see why this should require locking the session from down inside the query/control connection. The locking is non obvious and will add a burden to maintenance because now when the code changes the locking needs to be correctly move too. As well as being non obvious it will likely incur overall query performance degradation due to lock contention.

Zariel commented:
If it requires locking for concurrent metric updating then it should use much finer grained locking, per host/per query.

annismckenzie commented on session.go:
```diff
@@ -658,6 +658,11 @@ func (s *Session) connect(host *HostInfo, errorHandler ConnErrorHandler) (*Conn,
 }
 
+type queryMetrics struct {
+	attempts     int
```
That might be a bit much – one just has to remember that observers are being used with functions provided by the user. 😆

alourie replied:
maybe :-)

alourie commented:
@zariel @annismckenzie so I've cleared up all the things. Now not using locks at all, as indeed things run sequentially anyway. Will deal with locking when necessary in speculative query executions work.

Additionally:
* Exported the queryMetric fields so that an Observer can use them. 
* Internal test functions are updated so that it is possible to launch multinode mock cluster without providing different calls or slices of strings.
* Few subjective syntax improvements here and there
* Cleanups

alourie commented:
@annismckenzie do you mean the whole thing or just the last 3? I can squash it all  (or just the last ones), so that it is all just 1 commit.

annismckenzie commented:
Well, I see @Zariel usually squashes but that squash merge commit still links to this MR and I'd like it to be comprehensible. I really did mean to just do `git reset master` to do a soft reset to the current master state and just redoing the changes in 5 smaller logical commits:
- `Run Cassandra integration tests on 3.11.3 as well`
- `Add myself as a contributor`
- the code for creating multinode mock clusters in tests
- the whole `Record query metrics on a per-host basis` (could be argued that the tests should be their own commit)
- lastly, `Provide query observers with the per-host metrics`

I'd find that quite valuable and the squash commit merge is also nice and short.

alourie commented:
@annismckenzie uhm, you realize that what you're asking is quite bit of a manual work up with git's history, right? ;-) I don't mind doing it, it's just it will require force push, and while github will manage that fine, some comments/content may disappear. Is that ok with you?

annismckenzie commented:
Yup, exactly what I'm asking. You basically did 3 implementations here and while it's sometimes valuable to see the evolution of a feature through time that isn't the case here. We should strive to make the implementation you chose clear (it also helps with issues being reported to point to a specific commit that introduced it).

alourie commented:
@annismckenzie Done. Hope it is more clear this way.

alourie commented:
yay! Thanks @annismckenzie

Zariel commented on session.go:
```diff
@@ -713,13 +729,23 @@ func (q Query) String() string {
-	if q.attempts > 0 {
-		return q.totalLatency / int64(q.attempts)
+	attempts := 0
+	var latency int64 = 0
```
use latency := 0 or `var attempt int` dont mix both

alourie replied:
Done

Zariel commented on session.go:
```diff
@@ -1415,11 +1440,23 @@ func (s *Session) NewBatch(typ BatchType) *Batch {
+	hostMetrics, exists := b.metrics[host.ConnectAddress().String()]
+	if !exists {
+		// if the host is not in the map, it means it's been accessed for the first time
+		hostMetrics = &queryMetrics{Attempts: 0, TotalLatency: 0}
```
nit: go will initialize these values to 0 so you don't need to
